python/ray/train/checkpoint.py

```diff
@@ -0,0 +1,293 @@
+import contextlib
+import glob
+import json
+import logging
+import os
+import platform
+import shutil
+import tempfile
+import traceback
+from typing import Any, Dict, Iterator, List, Optional, Union
+import uuid
+
+import pyarrow.fs
+
+from ray.air._internal.filelock import TempFileLock
+from ray.train._internal.storage import _download_from_fs_path, _exists_at_fs_path
+from ray.util.annotations import PublicAPI
+
+logger = logging.getLogger(__name__)
+
+# The filename of the file that stores user metadata set on the checkpoint.
+_METADATA_FILE_NAME = ".metadata.json"
+
+# The prefix of the temp checkpoint directory that `to_directory` downloads to
+# on the local filesystem.
+_CHECKPOINT_TEMP_DIR_PREFIX = "checkpoint_tmp_"
+
+# This namespace UUID is an arbitrary constant needed to generate a UUID5 hash.
+_CHECKPOINT_UUID_NAMESPACE = uuid.UUID("627fe696-f135-436f-bc4b-bda0306e0181")
+
+
+@PublicAPI(stability="beta")
+class Checkpoint:
+    """A reference to data persisted as a directory in local or remote storage.
+
+    Access checkpoint contents locally using `checkpoint.to_directory()`.
+
+    Example creating a checkpoint using `Checkpoint.from_directory`:
+
+        >>> from ray.train.checkpoint import Checkpoint
+        >>> checkpoint = Checkpoint.from_directory("/tmp/example_checkpoint_dir")
+        >>> checkpoint.filesystem  # doctest: +ELLIPSIS
+        <pyarrow._fs.LocalFileSystem object...
+        >>> checkpoint.path
+        '/tmp/example_checkpoint_dir'
+
+    Example creating a checkpoint from a remote URI:
+
+        >>> checkpoint = Checkpoint("s3://bucket/path/to/checkpoint")
+        >>> checkpoint.filesystem  # doctest: +ELLIPSIS
+        <pyarrow._s3fs.S3FileSystem object...
+        >>> checkpoint.path
+        'bucket/path/to/checkpoint'
+
+    Attributes:
+        path: A path on the filesystem containing the checkpoint contents.
+        filesystem: PyArrow FileSystem that can be used to access data at the `path`.
+    """
+
+    def __init__(
+        self,
+        path: Union[str, os.PathLike],
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+    ):
+        """Construct a Checkpoint.
+
+        Args:
+            path: A local path or remote URI containing the checkpoint data.
+                If a filesystem is provided, then this path must NOT be a URI.
+                It should be a path on the filesystem with the prefix already stripped.
+            filesystem: PyArrow FileSystem to use to access data at the path.
+                If not specified, this is inferred from the URI scheme.
+        """
+        self.path = str(path)
+        self.filesystem = filesystem
+
+        if path and not filesystem:
+            self.filesystem, self.path = pyarrow.fs.FileSystem.from_uri(path)
+
+        # The UUID is generated by hashing the combination of the file system type
```

Comment:
I wonder if this could potentially be dangerous if the data was updated somehow. What if we made it purely randomly generated (presumably it gets carried along whenenever Checkpoint is passed to different workers within this class)?

Reply:
What about we just generate this `uuid` whenever `to_directory` gets called? That way we always use the latest path/filesystem rather than what it was at initialization.

If it's a random uuid, then we no longer de-duplicate downloads to the same directory.

Reply:
@ericl I think the random uuid idea also works in most use cases.

```python
ds.map_batches(Predictor, fn_args=(result.checkpoint,))  # <-- each map batches worker uses the same checkpoint w/ the same uuid

Trainer(resume_from_checkpoint=result.checkpoint)  # <-- each train worker downloads a ckpt with the same uuid
```

Only case it doesn't cover is multiple processes creating a separate checkpoints pointing to the same location.

I'm ok with either way, what do you think?

Reply:
Ok, let's go with the random uuid then, because I think this is what is currently implemented in the air.Checkpoint code.

Reply:
Oh, the `air.Checkpoint` currently uses a canonical uuid for URI-checkpoints (so same as the implementation I have now).

Reply:
Oh I see. Hmm, I feel it's a bit risky to use that so would still prefer to generate random ones to start at least.

Reply:
Ok, changed, ptal!